pallets/providers/src/lib.rs

```diff
@@ -0,0 +1,367 @@
+        traits::fungible::*,
+        Blake2_128Concat,
+    };
+    use frame_system::pallet_prelude::*;
+    use scale_info::prelude::fmt::Debug;
+
+    /// Configure the pallet by specifying the parameters and types on which it depends.
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Because this pallet emits events, it depends on the runtime's definition of an event.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// Type to access the Balances pallet (using the fungible trait from frame_support)
+        type NativeBalance: Inspect<Self::AccountId>
+            + Mutate<Self::AccountId>
+            + hold::Inspect<Self::AccountId>
+            // , Reason = Self::HoldReason> We will probably have to hold deposits
+            + hold::Mutate<Self::AccountId>
+            + freeze::Inspect<Self::AccountId>
+            + freeze::Mutate<Self::AccountId>;
+
+        /// Data type for the measurement of storage size
+        type StorageData: Parameter
+            + Member
+            + MaybeSerializeDeserialize
+            + Default
+            + MaybeDisplay
+            + AtLeast32BitUnsigned
+            + Copy
+            + MaxEncodedLen
+            + HasCompact;
+
+        /// Type that represents the total number of registered Storage Providers.
+        type UserCount: Parameter
+            + Member
+            + MaybeSerializeDeserialize
+            + Ord
+            + AtLeast32BitUnsigned
+            + FullCodec
+            + Copy
+            + Default
+            + Debug
+            + scale_info::TypeInfo
+            + MaxEncodedLen;
+
+        /// The minimum amount that an account has to deposit to become a storage provider.
+        #[pallet::constant]
+        type SpMinDeposit: Get<BalanceOf<Self>>;
+
+        /// The amount that a BSP receives as allocation of storage capacity when it deposits SpMinDeposit.
+        #[pallet::constant]
+        type SpMinCapacity: Get<StorageData<Self>>;
+
+        /// The slope of the collateral vs storage capacity curve. In other terms, how many tokens a Storage Provider should add as collateral to increase its storage capacity in one unit of StorageData.
+        #[pallet::constant]
+        type DepositPerData: Get<BalanceOf<Self>>;
+
+        /// The maximum amount of BSPs that can exist.
+        #[pallet::constant]
+        type MaxBsps: Get<u32>;
+
+        /// The maximum size of a multiaddress.
+        #[pallet::constant]
+        type MaxMultiAddressSize: Get<u32>;
+
+        /// The maximum number of protocols the MSP can support (at least within the runtime).
+        #[pallet::constant]
+        type MaxProtocols: Get<u32>;
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    // Storage:
+
+    /// The mapping from an account id to a main storage provider. Returns `None` if the account isnt a main storage provider.
```

Comment:
```suggestion
    /// The mapping from an account id to a main storage provider. Returns `None` if the account isn't a main storage provider.
```

Reply:
Also, it is not super important, but to try and keep consistency throughout the code, we write Main Storage Provider, or Backup Storage Provier (first letter in caps)